ethers-providers/src/rpc/transports/ws/manager.rs

```diff
@@ -0,0 +1,409 @@
+        if in_flight.method == "eth_subscribe" {
+            self.subs.service_subscription_request(id, in_flight.params.clone())?;
+        }
+
+        // Must come after self.subs.service_subscription_request. Do not re-order
+        tracing::debug!("Dispatching request to backend");
+        self.backend.dispatcher.unbounded_send(req).map_err(|_| WsClientError::DeadChannel)?;
+
+        self.reqs.insert(id, in_flight);
+        Ok(())
+    }
+
+    fn service_instruction(&mut self, instruction: Instruction) -> Result<(), WsClientError> {
+        match instruction {
+            Instruction::Request { method, params, sender } => {
+                let id = self.next_id();
+                self.service_request(id, method, params, sender)?;
+            }
+            Instruction::Unsubscribe { id } => {
+                if let Some(req) = self.subs.end_subscription(id.low_u64()) {
+                    self.backend
+                        .dispatcher
+                        .unbounded_send(req)
+                        .map_err(|_| WsClientError::DeadChannel)?;
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn spawn(mut self) {
+        let fut = async move {
+            let result = loop {
+                // We bias the loop so that we always handle messages before
+                // reconnecting, and always reconnect before dispatching new
+                // requests
+                select_biased! {
+                    item_opt = self.backend.to_handle.next() => {
+                        match item_opt {
+                            Some(item) => self.handle(item),
+                            // Backend is gone, so reconnect
+                            None => if let Err(e) = self.reconnect().await {
+                                break Err(e);
+                            }
+                        }
+                    },
+                    _ = &mut self.backend.error => {
+                        if let Err(e) = self.reconnect().await {
+                            break Err(e);
+                        }
+                    },
+                    inst_opt = self.instructions.next() => {
+                        match inst_opt {
+                            Some(instruction) => if let Err(e) = self.service_instruction(instruction) { break Err(e)},
+                            // User-facing side is gone, so just exit
+                            None => break Ok(()),
+                        }
+                    }
+                }
+            };
+            if let Err(err) = result {
+                tracing::error!(%err, "Error during reconnection");
+            }
+            // Issue the shutdown command. we don't care if it is received
+            self.backend.shutdown();
+        };
+
+        #[cfg(target_arch = "wasm32")]
+        super::spawn_local(fut);
+
+        #[cfg(not(target_arch = "wasm32"))]
+        tokio::spawn(fut);
+    }
+}
```

Comment:
nit: could add a `reconnects(..)` builder method to override the default, but can do in followup